openedx/openedx-platform
# Mark blocks completed on view in render_xblock view.

Webviews in mobile apps do not encapsulate leaf-blocks in html for vertical blocks. Instead, they use the "chromeless-courseware" `render_xblock` view. This PR adds this missing completion functionality to `render_xblock` view.

Part of [Completion API](https://openedx.atlassian.net/wiki/spaces/OpenDev/pages/162247762/Completion+API) work.
Previous PR: https://github.com/edx/edx-platform/pull/16234.

**JIRA tickets**: [OSPR-2441](https://openedx.atlassian.net/browse/OSPR-2441)

**Screenshots**: No change to UI.

**Sandbox URL**: https://pr18273.sandbox.opencraft.hosting/

**Merge deadline**: None

**Testing instructions**:

1. In `/admin/waffle/switch/` add and activate `completion.enable_completion_tracking` to turn on completion tracking.
2. Complete on view tracking happens when the user is visiting courseware pages and when the `render_xblock` view is displayed in the mobile apps.
3. To test courseware pages, visit a course section. When a vertical, which has XBlocks to which this completion strategy is applicable (e.g. html XBlock), is viewed the first time, a `publish_completion` ajax call is made to the server for each child XBlock when it has been viewed from start to end for 5 seconds.
4. To test `render_xblock` view, visit a url of the form `/xblock/<block_id>`. If this completion strategy is applicable to the XBlock, a `publish_completion` ajax call is made to the server when it is viewed for the first time from start to end for 5 seconds.
5. To help with testing, completion data can be reset via the following steps:
```
 ./manage.py lms shell
from completion.models import BlockCompletion
BlockCompletion.objects.all().delete()
```

**Reviewers**
- [x] @staubina or @iloveagent57
